src/modules/users-permissions/index.js

```diff
@@ -30,115 +30,137 @@ class UsersPermissions extends HasModels {
+      this.log.info("Created role", role)
+      roleObj.permissions = _.union(roleObj.permissions || [], this._defaultRoles[role])
+      return roleObj.save()
     })
   }
-  
+
+
   /*
    * Register a permission
    * @param {string} name Permission name
-   * @param {function|string} [handler|route] A handler or route to wrap with a permission check
-   * @param {object} [objectParams] Mapping of req params to ObjectRoleModel identity
-   * @param {string|array} [roleName] Default role name(s) to contain this permission
+   * @param {string|array} [defaultRoles] Default role name(s) to contain this permission
+   * @param {object} [objectModel] Collection of ObjectRoleModel for object permissions, or function (objectId, user) to return the roles a user has for a given object
    */
-  allow(name, handler = null, objectParams = null, roleName = null) {
+  register(name, defaultRoles=null, objectModel=null) {
+    this.log.info("Registering permission", name, "for", defaultRoles, objectModel)
     this._permissions[name] = {
       name,
-      handler,
-      objectParams,
+      objectModel,
     }
-    if(roleName) {
-      if (!_.isArray(roleName)) {
-        roleName = [roleName]
+    if(defaultRoles) {
+      if (!_.isArray(defaultRoles)) {
+        defaultRoles = [defaultRoles]
       }
-      for (let role of roleName) {
+      for (let role of defaultRoles) {
         if (!this._defaultRoles[role]) {
           this._defaultRoles[role] = []
         }
         this._defaultRoles[role].push(name)
       }
     }
+  }
+  
+  /*
+   * Guard a handler method or route a permission
+   * @param {function|string} [handler|route] A handler or route to wrap with a permission check
+   * @param {string} name Permission name
+   * @param {object} [objectParam] URL param for object permissions object key
+   * @returns {function} Wrapped handler for method handlers
+   */
+  guard(handler, name, objectParam=null) {
     if (_.isString(handler)) {
       // Route
-      this._routesPermissions.addRoute(handler, () => {return name})
+      this._routesPermissions.addRoute(handler, () => {return [name, objectParam]})
     } else {
       // Wrap handler
-      return this._checkPermission(name, handler)
+      return this._checkPermission(name, handler, objectParam)
     }
   }
+  /*
+   * Register a permission and guard a route handler in one
+   * @param {string} name Permission name
+   * @param {function|string} [handler|route] A handler or route to wrap with a permission check
+   * @param {object} [objectParams] Mapping of req params to ObjectRoleModel identity
+   * @param {string|array} [roleName] Default role name(s) to contain this permission
+   * @returns {function} Wrapped handler for method handlers
+   */
+  allow(name, handler = null, objectParams = null, roleName = null) {
+    let coll, param
+    if (objectParams) {
+      param = _.first(_.keys(objectParams))
```

Comment:
This seems rather magical, relying on the ordering of properties returned by `keys()` to pick name of the parameter to use. As I recall, that enumeration order depends on the order the properties were defined, which seems like it could lead to very obscure bugs.